### DISCUSSION TITLE
How can I read a global singleton property value from a secondary thread?

### DISCUSSION THREAD

**Discussion by matthewjrichey** [slint-ui/slint#5840] [Q&A]
I am developing a GUI app with Slint in the Rust language.

Is there any way, from native Rust code, to _read_ the value of a global property from a non-main or auxiliary thread? That is, from a thread that is not the one running the main GUI loop?

For example, let's say I create a top-level window called AppWindow from the main thread, where I have full access to the strong reference to the object. I then take a weak reference from that and make it accessible to a second thread. I can use that reference to, let's say, _set_ the value of a global property, but I cannot figure out how to *get* the property value.

Here is some code that fails at runtime:

```
//
// MAIN THREAD
//

// get a strong reference to the application window
let app_window = AppWindow::new().unwrap();

// create a weak reference from the strong reference
let app_window_weak = app_window.as_weak();

// spawn a new tokio runtime thread
let tokio_runtime = Runtime::new().unwrap();
slint::spawn_local(async move {
    tokio_runtime.spawn(async move{
        //
        // SECONDARY THREAD
        //

        // setting a property is fine because it executes in a closure
        // that runs in the context of the main thread
        let app_window_weak_clone = app_window_weak.clone();
        slint::invoke_from_event_loop(move || {
            app_window_weak_clone
                .unwrap()
                .global::<GlobalLogic>()
                .set_my_int_property(12345);
        });

        // however GET-ting a property cannot use invoke_from_event_loop()
        // because it only returns (), and unwrap()-ping the weak reference
        // from the secondary thread fails at runtime
        let my_value = app_window_weak
            .unwrap()
            .global::<GlobalLogic>()
            .get_my_int_property();
    })
    .await
    .unwrap();
})
.unwrap();
```

This code generates a runtime error:
```
thread 'tokio-runtime-worker' panicked at app_lib/src/lib.rs:105:14:
Trying to upgrade a Weak from a different thread than the one it belongs to
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at test_app/src/main.rs:23:10:
called `Result::unwrap()` on an `Err` value: JoinError::Panic(Id(11), ...)
```

Is there some other way to read/get() global singleton property values from a secondary thread? I could not find anything in the documentation.

Thank you for any help or suggestions.

**Comment by ogoffart** [comment]
Right now there is no easy way.
You can upgrade_in_event_loop. And in that callback send it. It would be nice if upgrade_in_event_loop returned a handle that can be awaited. But that'd be a breaking change (we would need to make a new function for that, I think)

**Reply by ogoffart** [reply]
In your case, I'd move the things inside the Tokio future by query ling it before. Or do something like

```rust

slint::spawn_local(async move {
    let xxx = tokio_runtime.spawn(async move{
        //
        // SECONDARY THREAD
        //
        xxx
     }).await;     
        
     let my_value = app_window
            .global::<GlobalLogic>()
            .get_my_int_property();
    });
})
```

**Comment by matthewjrichey** [comment]
Thanks for your reply.

Architecturally speaking, I am trying to figure out Slint's preferred way to maintain a single "source of truth" for the logical state of my application and to have this logical state automatically maintain parity with the GUI reactively.

One thought I had was to maintain the global source of truth in the global singleton properties themselves. Since the GUI already updates these automatically once the proper bindings are in place, then retrieving logical state should be as simple as invoking each property's `get_...()` method. However, as you've pointed out, this doesn't work from multiple threads, at least not in a straightforward way.

Another common GUI architecture strategy is to set up a state structure to mirror the GUI's properties, create a shared instance of it that is accessible to all threads (e.g. by wrapping it in something like `Arc<Mutex<...>>`), and have it maintain parity with the GUI by setting up the proper callbacks. But does Slint support a strategy like this? As far as I am aware, only *callbacks* have generated `on_()...` methods to register a native handler, but *properties* only have `get_...()` and `set_...()` methods.

How can I set up native code to react automatically to property updates? The only thing I can come up with after reading the Slint docs is to add a callback for each property that somehow ties in with a two-way binding, then setup native `on_...()` handlers for the callbacks. However, this seems like a lot of extra boilerplate to maintain.

Overall, what are Slint's best practices for implementing a reactive application architecture like what I've described here?

Thanks in advance for your help!
